File: src/fairseq2/data/tape.py
# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.
#
# This source code is licensed under the BSD-style license found in the
# LICENSE file in the root directory of this source tree.

__all__ = ["Tape"]

from typing import TYPE_CHECKING, final

from fairseq2 import DOC_MODE


@final
class Tape:
    def __init__(self) -> None:
        pass

    def rewind(self) -> None:
        pass


if not TYPE_CHECKING and not DOC_MODE:
    from fairseq2._C.data.tape import Tape  # noqa: F811

    Tape.__module__ = __name__

File: src/fairseq2/native/data/tape.cc
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
//
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

#include "fairseq2/native/data/tape.h"

namespace fairseq2 {

void
tape::record_data(const data &d)
{
    if (iter_ != storage_.end())
        throw std::logic_error{"New data can only be recorded to the end of the tape."};

    storage_.push_back(d);

    // At this point, the iterator is invalid because of the push_back call; do
    // not increment it.
    iter_ = storage_.end();
}

data
tape::read_data()
{
    if (iter_ == storage_.end())
        throw_corrupt();

    return *iter_++;
}

corrupt_tape_error::~corrupt_tape_error() = default;

}  // namespace fairseq2
